modules/loggers/mod_log_config.c

```diff
@@ -1602,13 +1634,204 @@ static ap_log_writer *ap_log_set_writer(ap_log_writer *handle)
+           code_point == 0x22 ||
+           code_point == 0x5c) {
+            strcats[isc].iov_base = apr_psprintf(p, "\\u%04x", code_point);
+            strcats[isc].iov_len = strlen(strcats[isc].iov_base);
+        } else {
+            strcats[isc].iov_base = (void*) utf8_string_to_encode + idx;
+            strcats[isc].iov_len = len;
+        }
+
+        isc++;
+        idx += len;
+        len = utf8_next_code_point(utf8_string_to_encode, idx, &code_point);
+    }
+
+    return apr_pstrcatv(p, strcats, isc, NULL);
+}
+
+static void add_iovec_str(struct iovec *iv, const char *str)
+{
+    iv->iov_base = (void*) str;
+    iv->iov_len = strlen(str);
+}
+
+static apr_status_t ap_json_log_writer( request_rec *r,
                            void *handle,
                            const char **strs,
                            int *strl,
                            int nelts,
+                           void *itms,
                            apr_size_t len)
 
+{
+    log_format_item *items = (log_format_item *) itms;
+    apr_size_t len_file_write;
+
+    const char* attribute_name;
+    const char* attribute_value;
+    const char* json_str;
+
+    struct iovec strcats[1 + (nelts * 7) ];
+    apr_size_t isc = 0;
+
+    // build json object
+    add_iovec_str(&strcats[isc++], "{");
+    for (int i = 0; i < nelts; ++i) {
+        if(items[i].tag == NULL) {
+                continue;
+        }
+
+        attribute_name = apr_hash_get(json_hash, items[i].tag, APR_HASH_KEY_STRING );
+        if(!attribute_name) {
+            attribute_name = items[i].tag; // use tag as attribute name as fallback
+
+            // TODO: do we really needs to check for json string encoding for tags?
+            if(json_needs_encoding(attribute_name)) {
+                attribute_name = json_encode(r->pool, attribute_name);
+            }
+        }
+
+        add_iovec_str(&strcats[isc++], "\"");
+        add_iovec_str(&strcats[isc++], attribute_name);
+
+        // process any arguemnts as attribute name extension
+        if(items[i].arg != NULL && strlen(items[i].arg) > 0) {
+            attribute_name = items[i].arg;
+            if(json_needs_encoding(attribute_name)) {
+                attribute_name = json_encode(r->pool, attribute_name);
+            }
+            add_iovec_str(&strcats[isc++], " ");
+            add_iovec_str(&strcats[isc++], attribute_name);
+        }
+        add_iovec_str(&strcats[isc++], "\":\"");
+
+        attribute_value = strs[i];
+        if(json_needs_encoding(attribute_value)) {
+            attribute_value = json_encode(r->pool, attribute_value);
+        }
+        add_iovec_str(&strcats[isc++], attribute_value);
+        add_iovec_str(&strcats[isc++], "\",");
+    }
+    // replace last '",' with '"}'
+    add_iovec_str(&strcats[isc - 1], "\"}" APR_EOL_STR);
+    json_str = apr_pstrcatv(r->pool, strcats, isc, &len_file_write);
+    return apr_file_write((apr_file_t*)handle, json_str, &len_file_write);
```

Comment:
Using `apr_file_writev_full()` directly with the iovec array would avoid allocating `json_str` here.